rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,223 @@
+
+A previous draft of this RFC would have required a detailed audit of the functionality of each app, to ensure we had Smoke or sandboxed E2E tests for each of its features, for some definition of "feature".
+
+This approach was rejected because:
+
+- There is no clear way to demarcate features of an app, let alone GOV.UK as a whole. Given the sheer volume of functionality in our apps, it's not practical to audit or have realistic tests for all of it.
+
+- While we have existing tests for specific features of GOV.UK [[1](https://github.com/alphagov/publishing-e2e-tests/tree/b4a7961cdcd200623a694baf833724a41c98d051/spec/collections_publisher)] [[2](https://github.com/alphagov/smokey/blob/a60df2cde5c886fa6d54e8cd820d90facca89e8e/features/csv_preview.feature)], there is no thorough, positive logic to justify their existence or extension [[1](https://github.com/alphagov/govuk-rfcs/blob/master/rfc-050-do-end-to-end-testing-of-gov-uk-applications.md#solution)] [[2](https://github.com/alphagov/publishing-e2e-tests/blob/f12ce7a9626b38728758cf157ea883907bd89de2/docs/what-belongs-in-these-tests.md)]. This is a weak foundation to try and build upon.
+
+## Consequences
+
+### Being more cautious
+
+We will no longer have the safety blanket of our changes languishing in Integration or Staging. It may take some time for us to get used to this new way of working e.g. by gradually doing more [branch deployments](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#review-your-own-changes).
+
+> **Quote from the trial of CD**: "We'll have to up our game to check things before deploying them. There are always corner cases you don't expect. We're perfectly capable of checking all our changes."
+
+### Automatic deployments
+
+We will enable automatic deployments for the apps listed [here](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/hieradata_aws/common.yaml#L145), but only when they meet the safety criteria of this RFC. To help with this, a rough audit of "compatibility" can be found in [Appendix A: Compatibility](rfc-128/compatibility.md).
+
+- We can immediately enable automatic deployments for apps marked with a ✅.
+- We will undertake work to address all the issues (⚠) for apps marked with a ❌.
+
+> **We will disable CD for Publishing API**. This was only enabled on a trial basis, and requires non-trivial work to bring it up to the new standard. We should avoid setting a false precedent by leaving CD enabled.
+
+The following steps must be taken as part of enabling automatic deployments:
+
+- Restrict merge access to "GOV.UK Production" [[example](https://github.com/alphagov/govuk-saas-config/pull/55)].
+
+- Tag [relevant Smoke tests](https://github.com/alphagov/smokey/blob/5dc36fc78c12838ec1c8070d6fbc535ae5433b89/docs/writing-tests.md#filtering-by-app) to run for the app [[example](https://github.com/alphagov/smokey/pull/708/commits/91a5ce94b7b96d09f5db87e63fbc3e400788de00)].
```

Comment:
I've also found the arbitrary organisation of Smokey quite irritating - it's hard to find some things. We can apply as many tags to a scenario as we like, so the organisation doesn't really hinder us.

Since this RFC makes relatively low usage of the Smokey tests, I do think it's out-of-scope to improve them. However, I do think we should make it clear what they are for. How about this:

### Check: App is running (i.e. `/healthcheck` returns 200 OK`)

(replaces existing section about Smokey)

### Check: Smoke tests pass (i.e. double check features work)

For the purpose of automating deployments:

- We do not strictly need to run any Smoke tests, as the combination of the other checks described here covers basically all the scenarios in which an app could break, due to a change in its repository.
- We should still check that existing Smoke tests for an app pass, as while they aren't strictly necessary, they act as a "backup" or "secondary" layer of testing for features that we have deemed to be critical.

... (consequences)

### Clarify Smokey

We will add documentation to [smokey](https://github.com/alphagov/smokey) to clarify how the tests it contains represent features that we have deemed to be critical to GOV.UK, and exist:

- For the purpose of testing changes to infrastructure.
- As a "backup" layer of testing for changes to individual apps.

Reply:
Yes. I like that suggestion, thanks

Reply:
Cool, I've added a commit for that 👍.